proposals/0000-fine-grained-unused-warnings.rst

```diff
@@ -0,0 +1,205 @@
+      ‘b1’ is defined but only used in the following unused bindings: ‘b2’, ‘b4’
+
+Related warnings that are not affected by this proposal are ``-Wunused-pattern-bindings``,
+``-Wunused-packages``, and ``-Wunused-do-bind``, since these are not about binding names.
+
+Examples
+--------
+
+NB: The only difference between these examples and how these warnings are currently presented is that the warnings about transitively unused bindings currently don't list the bindings in which they occur, and have the same phrasing as regular unused warnings - with the exception of the unused import warning and unused forall warnings, which don't currently occur if they are only transitively unused.
+
+::
+
+  module Foo () where
+
+  import Data.List as L
+
+  foo = L.intercalate bar
+
+  bar = baz
+    where baz = undefined
+          quux = wibble
+          wibble = worble
+          worble = quux
+        
+  far :: forall a (b :: a) c . c
+  far = far
+
+::
+
+  Foo.hs:3:1: warning: [-Wunused-imports, -freport-indirectly-used-bindings]:
+      The import of ‘Data.List’ is only used by the following unused binding: ‘foo’
+        except perhaps to import instances from ‘Data.List’
+      To import instances alone, use: import Data.List()
+
+  Foo.hs:5:1: warning: [-Wunused-top-binds]
+      Defined but not used: ‘foo’
+
+  Foo.hs:7:1: warning: [-Wunused-top-binds]
+      ‘bar' is defined but only used in the following unused binding: `foo’
+
+  Foo.hs:9:9: warning: [-Wunused-local-binds]
+      ‘quux' is defined but only used in the following unused bindings: `worble', `wirble’
+
+  Foo.hs:10:9: warning: [-Wunused-local-binds]
+      ‘wibble' is defined but only used in the following unused binding: `worble’
+
+  Foo.hs:11:9: warning: [-Wunused-local-binds]
+      ‘worble' is defined but only used in the following unused binding: `wibble’
+
+  Foo.hs:12:9: warning: [-Wunused-local-binds]
+      Defined but not used: ‘wirble’
+
+  Foo.hs:13:15: warning: [-Wunused-foralls]
+      Foo quantified type variable ‘a' is only used in the following unused variable: `(b :: a)’
+      In the type signature for ‘far’
+
+  Foo.hs:13:17: warning: [-Wunused-foralls]
+      Foo quantified type variable ‘(b :: a)’
+      In the type signature for ‘far’
+
+  Foo.hs:14:1: warning: [-Wunused-top-binds]
+      Defined but not used: ‘far’
+
+Effect and Interactions
+-----------------------
+For the most part, the effects of this proposal are minor. The only differences for existing
+code-bases are that the warning message GHC prints for the transitive warnings
+is different.
+
+Since the warnings don't have any special formats, existing tools should be able to handle them without issues.
+
+Users that don't wish to see warnings about transitively unused bindings can turn those warnings off.
+
+Costs and Drawbacks
+-------------------
+The warning mechanism is somewhat more complicated and as a consequence might
+have a somewhat higher maintenance cost.
```

Comment:
It is a bit of a shame that the new `-f` flag is somewhat separate from the `-W...` warning flag system, but changes which warnings are emitted. I wonder if there is a way to express this using warning groups (e.g. say that `-Wunused-imports` is split into `-Windirectly-unused-imports` + `-Wdirectly-unused-imports`, and have `-Wno-indirect-uses` to turn off all the "indirectly used" warnings at once). I haven't thought through all the consequences of this, though, and you may have considered it already.

Reply:
I've added it to alternatives. A point that came up earlier in discussion is people actually wanting less configurability, so from that point of view n + 1 flags (as in the current text) may be better than 2*n + 1 flags (or 3*n + 1, since we'd probably also want to keep the existing `-Wunused-imports`?). It's also not clear to me that having separate "indirectly unused" warnings for each category is something that would be used in practice.

Perhaps the `-f` flag in the text could simply be a `-W` name instead? Not sure if it'd be violating any implicit rules about how `-W` flags are supposed to work.